### PR TITLE
Hamdi/fee challenge

### LINKED ISSUE #84: Bug in StartExit
Currently the fee in a block is the aggregate FeeAmounts of all UTXOs included in that block.

So if a block contains UTXOs A, B, C:

FeeUTXO has `amount = Fee(A) + Fee(B) + Fee(C)`

However, if input to A decides not to send confirm sigs and exits on rootchain. Then FeeUTXO amount does not change, even if one of its inputs, A is not valid.

The correct FeeUTXO amount in this scenario would be:
FeeUTXO has `amount = Fee(B) + Fee(C)`

Thus with current implementation, anytime any input to an included UTXO is successfully exited, the Fee for that block has an incorrect amount that will inflate the number of Ether in the sidechain.

Possible Solutions:

1) Have users commit to any fee they spent in a transaction (whether included or not) during StartExit so that they can exit `UTXO.Amount - AlreadySpentFee`
Then we also need a way to challenge an Exit by proving that a spend of the UTXO was included on sidechain with a Fee that is different from the claimed fee by user during startExit. (Note: No confirm sigs are necessary for this particular challenge)

2) Possibly others, but I think they will be more complicated than the above fix

### PR DESCRIPTION
**Pull request by hamdiallam**
1.) Refactored libraries. Contract is too expensive to deploy - exceeds gas limit
2.) Fixed fee vulnerabilities. User must commit to fees payed when exiting.
3.) Renamed contract from RootChain -> PlasmaMVP

Fixes #84

### CODE REVIEW

File: contracts/PlasmaMVP.sol [hunk 1]
```diff
@@ -271,35 +291,62 @@ contract RootChain is Ownable {
...
             state: ExitState.Pending
         });
 
         // pass in empty bytes for confirmSignatures for StartedTransactionExit event.
-        emit StartedTransactionExit([blockNumber, txIndex, 0], msg.sender, feeAmount, "");
+        emit StartedTransactionExit([blockNumber, txIndex, 0], msg.sender, feeAmount, "", 0);
+    }
+
+    // @param exitedTxPos transaction position of the exit with an invalid committed fee
+    // @param challengingTxPos transaction posision of the inc
```
Comment (colin-axner) [inline contracts/PlasmaMVP.sol]: *position. Is "inc" a typo?

**Comment by colin-axner** [comment]
Add challenging for fee amount of deposits being exited.

File: docs/rootchainFunctions.md [hunk 2]
```diff
@@ -74,23 +75,23 @@ Note that if the validator attempts to start an exit for a fee-UTXO that has alr
...
-The `exitingTxPos` locates the malicious utxo and is used to calculate the priority. `challengingTxPos` locates the transaction that is the parent (offending transaction is an input into this tx).
-The `proof`, `txBytes` and `sigs` is sufficient for a proof of inclusion in the child chain of the parent transaction. The `confirmSignature`, signed by the owner of the malicious transaction,
-acknowledges the inclusion of it's parent in the plasma chain and allows anyone with this confirm signature to challenge a malicious exit of the child.
+An exit which posts an invalid committed fee can be challenged with this function. The `txBytes` of `challengingTxPos` which includes the correct fee, along with it's merkle `proof` of inclusion is checked against the exiter's claimed
+committed fee. If there is a mismatch, the exit is invalidated and the bond is awarded ot the challenger. `exitingTxPos` must be the first input of `challengingTxPos`. `exitingTxPos` is the full position including the deposit nonce.
```
Comment (colin-axner) [inline docs/rootchainFunctions.md]: *to

File: docs/rootchainFunctions.md [hunk 3]
```diff
@@ -74,23 +75,23 @@ Note that if the validator attempts to start an exit for a fee-UTXO that has alr
...
 
 <br />
 
 ```solidity
-function challengeDepositExit(uint256 nonce, uint256[3] newTxPos, bytes txBytes, bytes proof, bytes confirmSignature)
+function challengeExit(uint256[3] exitingTxPos, uint256[3] challengingTxPos, bytes txBytes, bytes proof, bytes confirmSignature)
```
Comment (colin-axner) [inline docs/rootchainFunctions.md]: update doc to *uint256[4] to match Plasma.sol

File: docs/rootchainFunctions.md [hunk 4]
```diff
@@ -74,23 +75,23 @@ Note that if the validator attempts to start an exit for a fee-UTXO that has alr
...
 ```
-A deposit that has been spent in the child chain is challenged here. The `txBytes` of the the parent transaction must include the nonce as one if it's input. The `txBytes`, `sigs` and `proof` is
-sufficient for a proof of inclusion. Similar to a normal challenge, the owner of the deposit must have also broadcasted a `confirmSignature` acknowledging the spend. A successfull challenge awards the
-caller with the exit bond.
+`challengingTxPos` follows the convention - `[blockNumber, transcationIndex, outputIndex]`
```
Comment (colin-axner) [inline docs/rootchainFunctions.md]: *transactionIndex

File: docs/rootchainFunctions.md [hunk 5]
```diff
@@ -74,23 +75,23 @@ Note that if the validator attempts to start an exit for a fee-UTXO that has alr
...
+`exitingTxPos` follows the convention - `[blockNumber, transactionIndex, outputIndex, depositNonce`]
+
+A uxto that has starting an exit phase but was already spent on the child chain can be challenged using this function call. A successfull challenge awards the caller with the exit bond.
```
Comment (colin-axner) [inline docs/rootchainFunctions.md]: *successful

File: docs/rootchainFunctions.md [hunk 6]
```diff
@@ -74,23 +75,23 @@ Note that if the validator attempts to start an exit for a fee-UTXO that has alr
...
+The `exitingTxPos` locates the malicious utxo and is used to calculate the priority. `challengingTxPos` locates the transaction that is the parent (offending transaction is an input into this tx).
```
Comment (colin-axner) [inline docs/rootchainFunctions.md]: `challengingTxPos` would be a child tx of the exitingTxPos right? Children should be younger than their parent

File: test/plasmamvp/deposits.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        let merkleHash = sha256String(txBytes);
+        let [merkleRoot, proof] = generateMerkleRootAndProof([merkleHash], 0);
+        let blockNum = (await instance.lastCommittedBlock.call()).toNumber() + 1;
+        await instance.submitBlock([toHex(merkleRoot)], [1], [5], blockNum, {from: authority});
+
+        // exit the deposit not committing to the fee
+        await instance.startDepositExit(nonce, 1, {from: accounts[2], value: minExitBond});
+
+        // challenge the exit
+        await instance.challengeFeeMismatch([0,0,0,nonce], [blockNum, 0, 0], toHex(txBytes), toHex(proof));
```
Comment (colin-axner) [inline test/plasmamvp/deposits.js]: can you add an err check here to ensure that it is successfully challenged

Comment (colin-axner) [reply test/plasmamvp/deposits.js]: Or a check that startDepositExit doesn't fail

Comment (hamdiallam) [reply test/plasmamvp/deposits.js]: added the check for the challenge. the deposit will revert if it doesn't succeed.

File: contracts/PlasmaMVP.sol [hunk 8]
```diff
@@ -87,34 +113,21 @@ contract RootChain is Ownable {
...
     // @param feesPerBlock amount of fees the validator has collected per block
     // @param blockNum     the block number of the first header
-    function submitBlock(bytes blocks, uint256[] txnsPerBlock, uint256[] feesPerBlock, uint256 blockNum)
+    function submitBlock(bytes32[] headers, uint256[] txnsPerBlock, uint256[] feesPerBlock, uint256 blockNum)
         public
-        onlyOwner
+        onlyOperator
     {
         require(blockNum == lastCommittedBlock + 1, "inconsistent block number ordering");
-        require(blocks.length > 0 && blocks.length % 32 == 0, "block roots must be of size 32 bytes");
```
Comment (colin-axner) [inline contracts/PlasmaMVP.sol]: do we need the length > 0 check? I could see some strange things occurring without it, thought I'm not 100% sure

Comment (colin-axner) [reply contracts/PlasmaMVP.sol]: actually nvm, assuming an empty array.length == 0 then it should be fine

File: contracts/PlasmaMVP.sol [hunk 9]
```diff
@@ -370,14 +385,14 @@ contract RootChain is Ownable {
...
 
         // reflect challenged state
-        txExits[position].state = ExitState.Challenged;
-        emit ChallengedExit(exit_.position, exit_.owner, exit_.amount);
+        exit_.state = ExitState.Challenged;
+        emit ChallengedExit(exit_.position, exit_.owner, exit_.amount - exit_.committedFee);
     }
 
     // When challenging an exiting transcation located at `exitingTxPos`, we must make sure that the challenging
     // transcation posted is either a direct spend of the exit if the confirm signature was not included in the txBytes of
```
Comment (colin-axner) [inline contracts/PlasmaMVP.sol]: "either" is not necessary. Delete part about confirm signature?

**Pull request merged**